onediff_comfy_nodes/utils/quant_ksampler_tools.py

```diff
@@ -0,0 +1,315 @@
+        out = latent_samples[0]
+        return out
+
+    def generate_pipeline(self, model_patcher, *args, **kwargs):
+        generate_img = self.generate_latent_sample
+
+        class ImageGenerationPipeline:
+            def __init__(self):
+                self.model = model_patcher
+                self.args = args
+                self.kwargs = kwargs
+
+            def __call__(self, *args, **kwargs):
+                return generate_img(self.model, *self.args, **self.kwargs)
+
+        return ImageGenerationPipeline()
+
+    def generate_calibrate_info(
+        self,
+        model_patcher,
+        only_compute_density: bool = False,
+        bits=8,
+        quantized_model_generator: callable = lambda x: x.model.diffusion_model,
+        resume: bool = True,
+        cached_process_output_path: str = None,
+        model_cls=[nn.Linear, nn.Conv2d],
+        *args,
+        **kwargs,
+    ) -> Dict:
+        """return calibrate_info"""
+
+        calibrate_info = {}
+        if resume and cached_process_output_path:
+            if os.path.exists(cached_process_output_path):
+                print(f"Resuming from {cached_process_output_path}")
+                calibrate_info = torch.load(cached_process_output_path)
+
+        diffusion_model = quantized_model_generator(model_patcher)
+
+        quantizable_modules = find_quantizable_modules(
+            diffusion_model, module_cls=model_cls
+        )
+
+        if set(quantizable_modules.keys()) == set(calibrate_info.keys()):
+            return calibrate_info
+
+        org_latent_sample = self.generate_latent_sample(model_patcher, *args, **kwargs)[
+            "samples"
+        ]
+        pipe = self.generate_pipeline(model_patcher, *args, **kwargs)
+        quantize_costs = metric_quantize_costs(
+            pipe, pipe_kwargs={}, quantizable_modules=quantizable_modules
+        )
+
+        if only_compute_density:
+            for sub_name, _ in quantizable_modules.items():
+                compute_density = quantize_costs.get_compute_density(sub_name)
+                calibrate_info[sub_name] = {
+                    "compute_density": compute_density,
+                }
+
+            torch.save(calibrate_info, cached_process_output_path)
+            return calibrate_info
+
+        length = len(quantizable_modules)
+        for index, (sub_name, _) in enumerate(quantizable_modules.items()):
+            print(f"Quantizing {index+1}/{length} {sub_name}...")
+            if sub_name in calibrate_info:
+                continue
+
+            with quantized_model_patcher(
+                model_patcher=model_patcher, layers=[sub_name], bits=bits
+            ) as qmpatcher:
+                cur_latent_sample = self.generate_latent_sample(
+                    qmpatcher, *args, **kwargs
+                )["samples"]
+
+                mse = torch.mean((org_latent_sample - cur_latent_sample) ** 2)
+
+                ssim = 1 - mse
```

Comment:
```suggestion
                similarity _mse= 1 - mse
```

SSIM 有专门的定义，为了避免混淆，给这个 `1-mse` 专门起个新名字吧。这样交流也不容易混淆（比如下面的报告，就可以直接 `ssim vs similarity_mse`。）